dpdk/vr_dpdk_virtio.c

```diff
@@ -798,12 +814,16 @@ vr_dpdk_virtio_enq_pkts_to_phys_lcore(struct vr_dpdk_queue *rx_queue,
 
     vq = (vr_dpdk_virtioq_t *) rx_queue->q_queue_h;
     vq_pring = vq->vdv_pring;
+    if (!vq_pring)
+        return;
 
     RTE_LOG(DEBUG, VROUTER, "%s: enqueue %u pakets to ring %p\n",
                 __func__, npkts, vq_pring);
     nb_enq = rte_ring_sp_enqueue_burst(vq_pring, (void **) pkt_arr, npkts);
-    for ( ; nb_enq < npkts; nb_enq++)
-        vr_pfree(pkt_arr[nb_enq], VP_DROP_INTERFACE_DROP);
+
+    for ( ; nb_enq < npkts; nb_enq++) {
+        vif_drop_pkt(pkt_arr[nb_enq]->vp_if, pkt_arr[nb_enq], 0, VP_DROP_ENQUEUE_FAIL);
```

Comment:
> which is for TX from the VM
Isn't it RX? The comment above the function declaration says: 'enqueue packets received on a virtio interface queue onto a ring that will be handled by the lcore'.

>and the enqueue failure when we enqueue to a ring for packets being sent to the VM?
We do it in two places: dpdk_virtio_to_vm_tx() - here packets are dropped when queue is not ready to accept packets, so I consider this as ENQUEUE_FAIL. We want to send, TX queue does not cooperate, so drop. According to that, I would not change the error code.
dpdk_virtio_to_vm_flush() - here packets are processed before sending to the VM. They were dequeued somewhere else, and if something fails during processing (no address, no place in buffer), they're dropped as DEQUEUE_FAIL, so with different error code. This is what I understood from our previous email exchange on how the queues work.